### PR TITLE
virt-handler: run as non-root user

### LINKED ISSUE #113: Don't run our containers with root user
The following containers should not be run as root:

 - [x] virt-api (#257)
 - [x] haproxy (#270) 
 - [x] virt-controller (#271) 
 - [x] squid-proxy (#282)
 - [ ] virt-handler (this container is privileged, let's see if we can then still talk to libvirt if we are not root)

### PR DESCRIPTION
**Pull request by cynepco3hahue**
Part of the fix #113 
Signed-off-by: Lukianov Artyom <alukiano@redhat.com>

### CODE REVIEW

File: cmd/virt-handler/Dockerfile [hunk 1]
```diff
@@ -28,4 +28,9 @@ RUN yum -y install wget && \
 
 COPY virt-handler /virt-handler
 
+RUN groupadd -g 993 libvirt
```
Comment (fabiand) [inline cmd/virt-handler/Dockerfile]: How did you come up with a uid for this user? and why this uid?

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: We have to explicitly set the gid and uid in the libvirt image we are using. Then we can synchronize them.

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: https://github.com/kubevirt/libvirt/issues/5

Comment (berrange) [inline cmd/virt-handler/Dockerfile]: Relying on matching UIDs across different containers is not very attractive IMHO and does little to nothing for security.

File: images/libvirt-kubevirt/Dockerfile [hunk 2]
```diff
@@ -44,4 +44,12 @@ RUN rm -f /libvirtd.sh
 COPY libvirtd.sh /libvirtd.sh
 RUN chmod a+x /libvirtd.sh
 
+RUN echo "unix_sock_group=\"libvirt\"" >> /etc/libvirt/libvirtd.conf
+RUN echo "unix_sock_ro_perms=\"0777\"" >> /etc/libvirt/libvirtd.conf
+RUN echo "unix_sock_rw_perms=\"0770\"" >> /etc/libvirt/libvirtd.conf
+RUN echo "auth_unix_ro=\"none\"" >> /etc/libvirt/libvirtd.conf
+RUN echo "auth_unix_rw=\"none\"" >> /etc/libvirt/libvirtd.conf
```
Comment (berrange) [inline images/libvirt-kubevirt/Dockerfile]: Enable 'sasl' auth here instead, and set the 'scram-sha-1' mechanism in the SASL config. Then we can setup a username + password for libvirtd & provide that via an libvirt auth config file to virt-handler.

File: cmd/virt-handler/Dockerfile (hunk 1 above)
Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: Created https://github.com/kubevirt/libvirt/issues/6 based on your feedback.

File: images/libvirt-kubevirt/Dockerfile (hunk 2 above)
Comment (cynepco3hahue) [inline images/libvirt-kubevirt/Dockerfile]: Thanks for comments guys, I checked the possibility to change auth to SASL and it will work, but we will need to explicitly send the same user and password that we used for `saslpasswd2` command under calls to `virtwrap.NewConnection`.
I believe we can define some default values for user and password under `config.sh` and use it when we call to the container, something like
```
command:
          - "/virt-handler"
          - "-v"
          - "3"
          - "--libvirt-uri"
          - "qemu:///system"
          - "--hostname-override"
          - "$(NODE_NAME)"
          - {{ sasl_user }}
          - {{ sasl_password }}
```
@rmohr @berrange @fabiand Give me know if it looks fine for you.

Comment (rmohr) [inline images/libvirt-kubevirt/Dockerfile]: @cynepco3hahue I think what @berrange had in mind, is to supply a client configuration. Then you don't have to provide the password and the user via commandline: https://libvirt.org/auth.html#Auth_client_config

Comment (cynepco3hahue) [inline images/libvirt-kubevirt/Dockerfile]: @rmohr I do not sure that go binding for libvirt will check auth.conf at all, because we use `libvirt.NewConnectWithAuth` in our `newConnection` method and this method assume that you will give to it user and password

Comment (berrange) [inline images/libvirt-kubevirt/Dockerfile]: You must never provide passwords as command line arguments or env variables as this is insecure. Libvirt has ability to use a client config file for providing passwords as described by that link rmohr provides. The Go binding has no involvement in this process - it is all handled by the underlying libvirt library that Go calls.

Comment (cynepco3hahue) [inline images/libvirt-kubevirt/Dockerfile]: @berrange Thanks for the clarification, but anyway we will need to hardcode sasl user and password for virt-manifest and virt-handler containers in the `auth.conf` file and also under saslpasswd2 command.
Can you give me some tips how can I avoid it?

Comment (berrange) [inline images/libvirt-kubevirt/Dockerfile]: We should be using a k8s 'secret' object to load the credentials into kubernetes, and then an Docker image entry script to setup the config / run saslpasswd2

Comment (rmohr) [inline images/libvirt-kubevirt/Dockerfile]: @cynepco3hahue We can use Kubernetes secrets (https://kubernetes.io/docs/concepts/configuration/secret/) to provide the credentials. A container can find the decrypted secrets in a special file, which is mounted by Kubernetes. To use the credentials, we have to modify our start-up scripts for libvirt and virt-handler. The libvirt start-script has to take the secrets from the file and place them in the server config, for virt-handler we have to place them in the client config. By doing this on every container start, we don't have to hardcode them.

Comment (cynepco3hahue) [inline images/libvirt-kubevirt/Dockerfile]: @rmohr @berrange Thanks guys, will check it

File: cmd/virt-handler/Dockerfile [hunk 3]
```diff
@@ -23,9 +23,20 @@ FROM centos:7
...
 RUN yum -y install wget && \
     wget -P /etc/yum.repos.d https://copr.fedorainfracloud.org/coprs/jmliger/virt7-upstream/repo/epel-7/jmliger-virt7-upstream-epel-7.repo  \
-    && yum -y install libvirt-client \
+    && yum -y install libvirt-client cyrus-sasl-md5 \
     && yum -y clean all
 
+COPY auth.conf /etc/libvirt/auth.conf
 COPY virt-handler /virt-handler
+COPY virt-handler.sh /virt-handler.sh
+RUN chmod a+x /virt-handler.sh
```
Comment (davidvossel) [inline cmd/virt-handler/Dockerfile]: Every one of these RUN commands creates a separate layer. By combining them into a single RUN execution, we can collapse all of these layers into one.  Reducing layers helps with performance.

```
RUN cmd1 && \
   cmd2 && 
   cmd3 && \
   ect...
```

File: images/libvirt-kubevirt/Dockerfile [hunk 4]
```diff
@@ -39,6 +39,13 @@ RUN chmod 0640 /etc/sudoers.d/kubevirt
 COPY libvirtd-limited.sh /libvirtd-limited.sh
 RUN chmod a+x /libvirtd-limited.sh
 
+# Configure SASL
+RUN echo "auth_unix_rw=\"sasl\"" >> /etc/libvirt/libvirtd.conf
```
Comment (davidvossel) [inline images/libvirt-kubevirt/Dockerfile]: see other comment.  combine RUN when possible

File: cmd/virt-handler/Dockerfile [hunk 5]
```diff
@@ -23,9 +23,20 @@ FROM centos:7
 # no-lower than what was used when the binary was built.
 RUN yum -y install wget && \
     wget -P /etc/yum.repos.d https://copr.fedorainfracloud.org/coprs/jmliger/virt7-upstream/repo/epel-7/jmliger-virt7-upstream-epel-7.repo  \
-    && yum -y install libvirt-client \
+    && yum -y install libvirt-client cyrus-sasl-md5 \
```
Comment (berrange) [inline cmd/virt-handler/Dockerfile]: Don't use cyrus-sasl-md5 - that is deprecated - use  cyrus-sasl-scram

File: cmd/virt-manifest/Dockerfile [hunk 6]
```diff
@@ -23,9 +23,12 @@ FROM centos:7
 # no-lower than what was used when the binary was built.
 RUN yum -y install wget && \
     wget -P /etc/yum.repos.d https://copr.fedorainfracloud.org/coprs/jmliger/virt7-upstream/repo/epel-7/jmliger-virt7-upstream-epel-7.repo  \
-    && yum -y install libvirt-client \
+    && yum -y install libvirt-client cyrus-sasl-md5 \
```
Comment (berrange) [inline cmd/virt-manifest/Dockerfile]: Same comment as above

File: manifests/sasl-sercret.yaml.in [hunk 7]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: sasl-secret
+type: Opaque
+data:
+  username: bGlidmlydC1jbGllbnQ=
+  password: Mm1mS1NGNXA=
```
Comment (berrange) [inline manifests/sasl-sercret.yaml.in]: You shouldn't hardcode a password. 

First add a "libvirt_password" env variable to hack/config.sh (initializing it to a randomly generated string)

Make hack/build-manifests.sh know about this new env variable for subsitution

Finally use  'password: {{ libvirt_password }}' here.

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (cynepco3hahue) [inline cmd/virt-handler/Dockerfile]: I tried to install on libvirtd and also on virt-handler containers cyrus-sasl-scram package only, but when I tried to connect to the libvirt I received the error:
`Failed to start SASL negotiation: -4 (SASL (-4) : no mechanism available: No worthy mechs found)`

**Comment by cynepco3hahue** [comment]
This tests periodically falls(passed locally), can we open a issue for this one
Failure in Spec Teardown (AfterEach) [0.001 seconds]
Console
/home/travis/gopath/src/kubevirt.io/kubevirt/pkg/virt-handler/rest/console_test.go:180
  with existing domain [AfterEach]
  /home/travis/gopath/src/kubevirt.io/kubevirt/pkg/virt-handler/rest/console_test.go:175
    should proxy websocket traffic
    /home/travis/gopath/src/kubevirt.io/kubevirt/pkg/virt-handler/rest/console_test.go:173
    missing call(s) to *virtwrap.MockVirDomain.Free()

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (michalskrivanek) [inline cmd/virt-handler/Dockerfile]: @cynepco3hahue, hm, why are you adding libvirt auth anyway? Is it needed for virt-handler to run as non-root?

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: The socket can be seen from the host, and other pods could mount the host directory with the socket and talk to libvirt. We can't influence what non-KubeVirt workloads are doing.

Comment (michalskrivanek) [inline cmd/virt-handler/Dockerfile]: sure, that's worth summing up since we have some experience using the "secret" password to protect ourselves from BFUs fiddling with internals:), but how is that related to this particular PR?

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: Because, if we run virt-handler as non-root, there is no guarantee, that uids/gids match between the libvirt and the virt-handler container. Therefore we either would have to synchronize the uids/gids, to not expose the socket read-and writeable for anyone. That however is still insecure, since any other container can use the same uids/gids. Therefore, we expose the socket now read/write-able for anyone, but protect it by a password, which is delivered via a kubernetes secret. This way we don't have to synchronize gids/uids and have a more secure solution.

File: images/libvirt-kubevirt/libvirtd.sh [hunk 8]
```diff
@@ -80,6 +80,11 @@ fi
 
 echo "cgroup_controllers = [ ]" >> /etc/libvirt/qemu.conf
 
+# Configure libvirt SASL authentication
+echo "auth_unix_rw=\"sasl\"" >> /etc/libvirt/libvirtd.conf
+echo "mech_list: scram-sha-1" >> /etc/sasl2/libvirt.conf
+cat /etc/sasl-secret/password | saslpasswd2 -p -c -a libvirt $(cat /etc/sasl-secret/username)
+
 if [[ -n "$LIBVIRTD_DISABLE_TCP" ]]; then
```
Comment (michalskrivanek) [inline images/libvirt-kubevirt/libvirtd.sh]: since we're now touching libvirtd.conf - why not to configure TCP there instead?

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (michalskrivanek) [inline cmd/virt-handler/Dockerfile]: sounds ok
for debugging purposes - what would be the best way to get in?

File: cmd/virt-handler/virt-handler.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/bash
+
+# Configure libvirt username and password
+sed -i s/^authname=/authname=$(cat /etc/sasl-secret/username)/g /etc/libvirt/auth.conf
+sed -i s/^password=/password=$(cat /etc/sasl-secret/password)/g /etc/libvirt/auth.conf
+
+# Start virt-handler
+/virt-handler -v 3 --libvirt-uri qemu:///system --hostname-override $NODE_NAME
```
Comment (michalskrivanek) [inline cmd/virt-handler/virt-handler.sh]: why supply the parameters? we didn't do that prior to this PR, did we?

File: cmd/virt-handler/virt-handler.sh [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/bash
+
+# Configure libvirt username and password
+sed -i s/^authname=/authname=$(cat /etc/sasl-secret/username)/g /etc/libvirt/auth.conf
+sed -i s/^password=/password=$(cat /etc/sasl-secret/password)/g /etc/libvirt/auth.conf
```
Comment (michalskrivanek) [inline cmd/virt-handler/virt-handler.sh]: would be easier to just generate that file.
Or virt-handler can use it natively if you keep(and enhance) the "--libvirt-uri" parameter, then no need for these files....

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: The username and secret are mounted inside the libvirt and the virt-handler container. The simplest solution, is to `kubectl exec` on a running virt-handler container. It already contains a properly configured client auth file. Just running `virsh` in there will work, just like it does right now, without authentication.

An example, how to view VMs:

```
kubectl exec virt-handler-z0ctc -- virsh list
 Id    Name                           State
----------------------------------------------------
 1     testns_testvm                  running
```

Using virsh interactive works like this:
```
kubectl exec -it virt-handler-z0ctc -- virsh
Welcome to virsh, the virtualization interactive terminal.

Type:  'help' for help with commands
       'quit' to quit

virsh # list                                                                                                                                                                            
 Id    Name                           State
----------------------------------------------------
 1     testns_testvm                  running

virsh # 
```

Comment (michalskrivanek) [inline cmd/virt-handler/Dockerfile]: sounds ok, though creating a seamless configuration in the libvirt container would be more intuitive, even if just for the sake of troubleshooting

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: That will work when we also create the auth file in the libvirt container. @cynepco3hahue maybe we want to do that?

File: images/libvirt-kubevirt/libvirtd.sh (hunk 8 above)
Comment (cynepco3hahue) [inline images/libvirt-kubevirt/libvirtd.sh]: I do not have a problem to do it, but does it really matter in particular case(do it have some advantage over regular Unix sockets?)

File: cmd/virt-handler/virt-handler.sh (hunk 9 above)
Comment (cynepco3hahue) [inline cmd/virt-handler/virt-handler.sh]: No, it does not connect to the PR, I just took it from the Dockerfile

File: cmd/virt-handler/virt-handler.sh (hunk 10 above)
Comment (cynepco3hahue) [inline cmd/virt-handler/virt-handler.sh]: Like @berrange said in comments above it does not secure to provide the password via command line arguments, but I can just to generate auth.file instead of `COPY` and `sed` commands.

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (cynepco3hahue) [inline cmd/virt-handler/Dockerfile]: Just to be sure that I understand correctly, do you want to create the auth file in the libvirt container too?

File: images/libvirt-kubevirt/libvirtd.sh (hunk 8 above)
Comment (berrange) [inline images/libvirt-kubevirt/libvirtd.sh]: Use of the TCP socket is insecure unless you can enable Kerberos, which we do not want. UNIX socket or TLS socket are the only options and for local connections it is preferred to use UNIX socket for performance reasons.

File: cmd/virt-handler/Dockerfile [hunk 11]
```diff
@@ -23,9 +23,20 @@ FROM centos:7
...
+COPY auth.conf /etc/libvirt/auth.conf
 COPY virt-handler /virt-handler
+COPY virt-handler.sh /virt-handler.sh
+RUN chmod a+x /virt-handler.sh
 
-ENTRYPOINT [ "/virt-handler" ]
+RUN useradd -u 1000 --create-home -s /bin/bash virt-handler && \
+    chown virt-handler /etc/libvirt && \
+    chown virt-handler /etc/libvirt/auth.conf && \
+    chown virt-handler /virt-handler.sh
```
Comment (berrange) [inline cmd/virt-handler/Dockerfile]: Ewww, should not be changing any of these file owners. Leave them all as their normal user. The auth.conf file should be in $HOME/.config/libvirt/auth.conf of the virt-handler user account, not /etc/libvirt

File: cmd/virt-handler/virt-handler.sh (hunk 9 above)
Comment (berrange) [inline cmd/virt-handler/virt-handler.sh]: Don't put these parameters here - leave them in the k8s manifest, as we need to be able to customize tem without rebuilding the image.

File: cmd/virt-handler/Dockerfile (hunk 5 above)
Comment (berrange) [inline cmd/virt-handler/Dockerfile]: Yes, all the server side auth setup must be in the libvirtd container - the virt-handler container merely needs to have an auth.conf file added to let to connect.

**Comment by cynepco3hahue** [comment]
@berrange Hi Daniel, maybe you can help me, after rebase each time that I try to connect to libvirt via
`# virsh -c qemu:///system` I receive the error `error: failed to connect to the hypervisor
error: authentication failed: authentication failed`
and under the log I can see:
`2017-07-25 13:27:17.294+0000: 18477: error : virNetSASLSessionListMechanisms:390 : internal error: cannot list SASL mechanisms -4 (SASL(-4): no mechanism available: Internal Error -4 in server.c near line 1757)
2017-07-25 13:27:17.294+0000: 18477: error : remoteDispatchAuthSaslInit:3415 : authentication failed: authentication failed
2017-07-25 13:27:17.294+0000: 18465: error : virNetSocketReadWire:1808 : End of file while reading data: Input/output error`.
Maybe you know what can be the problem?

**Comment by berrange** [comment]
The error message suggests you've not installed any sasl mechanisms

**Comment by cynepco3hahue** [comment]
@berrange I installed all cyrus packages that I had via repository
```
# rpm -qa | grep cyrus
cyrus-sasl-gssapi-2.1.26-32.fc26.x86_64
cyrus-imapd-vzic-3.0.2-4.fc26.x86_64
qca-qt5-cyrus-sasl-2.1.3-6.fc26.x86_64
cyrus-sasl-ldap-2.1.26-32.fc26.x86_64
cyrus-sasl-2.1.26-32.fc26.x86_64
cyrus-sasl-lib-2.1.26-32.fc26.x86_64
cyrus-imapd-3.0.2-4.fc26.x86_64
cyrus-imapd-devel-3.0.2-4.fc26.x86_64
cyrus-sasl-devel-2.1.26-32.fc26.x86_64
cyrus-sasl-sql-2.1.26-32.fc26.x86_64
qca-cyrus-sasl-2.1.3-6.fc26.x86_64
cyrus-sasl-md5-2.1.26-32.fc26.x86_64
cyrus-sasl-ntlm-2.1.26-32.fc26.x86_64
cyrus-sasl-gs2-2.1.26-32.fc26.x86_64
cyrus-sasl-scram-2.1.26-32.fc26.x86_64
cyrus-imapd-utils-3.0.2-4.fc26.x86_64
lua-cyrussasl-1.1.0-2.fc26.x86_64
cyrus-sasl-plain-2.1.26-32.fc26.x86_64
```

**Comment by cynepco3hahue** [comment]
@berrange Also in case when I did not have SASL mechanism I had the different error
`Failed to start SASL negotiation: -4 (SASL (-4) : no mechanism available: No worthy mechs found`

**Comment by berrange** [comment]
Then perhaps you're missing the sasl config file for libvirt to select which to use.

**Comment by cynepco3hahue** [comment]
@berrange Thanks for help, I re-checked `/etc/sasl2/libvirt.conf` and saw that I have additional `mech_list:` variable that not commented by default

**Comment by rmohr** [comment]
@cynepco3hahue could you rebase?

**Comment by cynepco3hahue** [comment]
@rmohr Does it need rebase?

**Comment by rmohr** [comment]
It's just to have the latest test script. Functional test runs are enabled again, and they need this script.

**Comment by rmohr** [comment]
ok to test

File: cmd/virt-handler/Dockerfile [hunk 12]
```diff
@@ -20,9 +20,18 @@ FROM fedora:26
 
 MAINTAINER "The KubeVirt Project" <kubevirt-dev@googlegroups.com>
 
-RUN dnf -y install libvirt-client \
+RUN dnf -y install libvirt-client cyrus-sasl-scram \
```
Comment (admiyo) [inline cmd/virt-handler/Dockerfile]: So, just invastigated scram and it looks like it only does sha1, which is weak.  Is there another option?

Comment (cynepco3hahue) [inline cmd/virt-handler/Dockerfile]: I believe the scram is the single option if we do not want to over complex authentification with the kerberos or another certification mechanism. Also, virt-handler talk with libvirt via unix socket on the same host, so I believe scram is more than enough.

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: I don't know how far everything is, regarding to [scram-sha-256](https://tools.ietf.org/html/rfc7677), but right now, I think it is good enough, since it is the safest option we have and definitely way better than the unprotected socket before. @admiyo what do you say?

Comment (rmohr) [inline cmd/virt-handler/Dockerfile]: @admiyo ping?

File: cmd/virt-handler/Dockerfile [hunk 13]
```diff
@@ -20,9 +20,18 @@ FROM fedora:26
...
     && dnf -y clean all
 
 COPY virt-handler /virt-handler
+COPY virt-handler.sh /virt-handler.sh
+RUN chmod a+x /virt-handler.sh
```
Comment (davidvossel) [inline cmd/virt-handler/Dockerfile]: Combine these two RUN lines.  Less layers in the union mount filesystem == better performance.

File: manifests/virt-handler.yaml.in [hunk 14]
```diff
@@ -26,9 +25,13 @@ spec:
           - "$(NODE_NAME)"
         securityContext:
           privileged: true
+          runAsNonRoot: true
```
Comment (davidvossel) [inline manifests/virt-handler.yaml.in]: The point is not to run as root... right?!

Comment (cynepco3hahue) [inline manifests/virt-handler.yaml.in]: runAsNonRoot?

File: cmd/virt-handler/Dockerfile (hunk 13 above)
Comment (cynepco3hahue) [inline cmd/virt-handler/Dockerfile]: Will merge this two lines

File: manifests/virt-handler.yaml.in (hunk 14 above)
Comment (davidvossel) [inline manifests/virt-handler.yaml.in]: ha, ignore this, I misread the flag's name :)

**Comment by cynepco3hahue** [comment]
In the beginning of the patch I already tried this approach, but I just used libvirt group and did not think to create absolutely new groups on the libvirt and virt-handler containers. But I already spent a lot of time on this patch and now we have secure enough connection between libvirt and virt-handler containers, so I prefer to leave it as it now.

**Comment by davidvossel** [comment]
>  But I already spent a lot of time on this patch and now we have secure enough connection between libvirt and virt-handler containers, so I prefer to leave it as it now.

I know this is a bit frustrating. The problem is the method you have now involves a lot of moving pieces (sharing secrets, file manipulation, etc..) and it involves introducing new dependencies (cyrus-sasl-scram).

If we have to do all those things to get this to work, that's fine. We need to be certain we've exhausted other options first. 

What was the problem you encountered the first round when you attempted to use the libvirt group?

**Comment by michalskrivanek** [comment]
Hmm, not really related to this PR, but I still keep wondering about libvirt on host. 
We need VM monitoring via libvirt, and doing that separately inside each POD is not really an option due to increased complexity and performance problems - we need libvirt bulk calls. I'm not sure how we can do that with libvirt in each POD. virt-handler doesn't need that, but someone does have to do it, and it will need visibility to all VMs on host
Thoughts?

**Comment by rmohr** [comment]
@davidvossel 

> What was the problem you encountered the first round when you attempted to use the libvirt group?

The problem with the groups is, that every container can randomly choose its uids and guids, and when they mount in the socket directory, they can just talk to the socket. Further, we have to synchronize guids and uids over container boundaries. While this one is not that simple, I think it is simple enough on one hand and secure enough on the other hand, that we just don't have to think about this anymore. We can just make the socket accessible to everyone who has the secret.

**Comment by rmohr** [comment]
@michalskrivanek 

> Hmm, not really related to this PR, but I still keep wondering about libvirt on host.
> We need VM monitoring via libvirt, and doing that separately inside each POD is not really an option due to increased complexity and performance problems - we need libvirt bulk calls. I'm not sure how we can do that with libvirt in each POD. virt-handler doesn't need that, but someone does have to do it, and it will need visibility to all VMs on host
> Thoughts?

I am not sure the mentioned problem exists. We just have one libvirt per host. So the complexity should be the same as always, just that you connect to the one libvirt, running inside the pod.

**Comment by davidvossel** [comment]
> The problem with the groups is, that every container can randomly choose its uids and guids, and when they mount in the socket directory, they can just talk to the socket. Further, we have to synchronize guids and uids over container boundaries.

It's pretty trivial to sync uids and gids in the docker build step.

Example: create user and group with a specific ids
```
RUN groupadd --gid 1111 somegroup && useradd --uid 1111 --gid 1111 someuser
```

Example: modify an existing user/group to match specific id
```
RUN groupmod -g 1111 somegroup && usermod  -u 1111 someuser
```
That seems way more simple IMO than having to randomly generate a libvirt secret and inject that secret into all the manifests.

Is there a reason why the secret option is any more secure?

**Comment by rmohr** [comment]
There are a few cumbersome side-effects (which I was willing to take, but this here is imho better).

* You either have to create the groups/users at container build time (which means that you build in config values), or
* you have to add them via starting as root user, and dynamically add them (which requires a container which can run as root)

with that, every other container could still just add the same user, mount in the directory and access your socket. 

So it is not secure, and you have to modify the docker containers already. Secrets can solve that for us. They are at the end part of the `kubevirt` namespace, where only admins have access too. So tenant users can't reference the secret, and they could even mount in the socket folder without causing issues. I agree that it is a little bit more work, but it is imho worth it.

**Comment by michalskrivanek** [comment]
@rmohr , ech, ok, I guess I just got confused by all those discussions, and also the diagram https://github.com/kubevirt/kubevirt/blob/master/docs/architecture.png which clearly depicts libvirtd in a POD;-) 

@rmohr, what's the problem of building the hardcoded UID/GID at build time? I'm not sure what's supposed to be more secure the way it's proposed in this PR? Socket permissions are by far more easy to handle and IMO it doesn't make it any more or less secure. I agree with @davidvossel it's just too complicated for the simple problem, and on top of that I do not really like one more hurdle to overcome to troubleshoot libvirt(get the secret or proxy through existing POD)

**Comment by cynepco3hahue** [comment]
@michalskrivanek Node can run additional container that does not relate to KubeVirt with root permissions, and this container can easily get access to VM container via libvirt that runs on the same node and SASL authentification will add additional security layer that will prevent it. I agree it does not relate to the PR subject, but I believe we will need in the near future to think about security, anyway so it can be a good start, IMHO.

**Comment by michalskrivanek** [comment]
making sure tenants or other random "users" of the host cannot access libvirt is not a bad idea. We need to solve that not only for the socket but the rest of host resources exposed to the libvirt container. I wonder what are our options...
Either way, it's separate from this PR I guess.

**Comment by rmohr** [comment]
We can use [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) (look at the volume section). There we can e.g. not allow users to mount host directories, to further improve security. After discussing this with @michalskrivanek in more details, I also wonder, if securing the socket this way really gives us more security, or just the feeling  to be more secure.

As @cynepco3hahue said the main reason for this PR was, to allow us to run the container unprivileged, not to secure libvirt. With scram authentication in place, it is probably a little bit more secure, but I leave the final decision to you @davidvossel and @cynepco3hahue .

**Comment by cynepco3hahue** [comment]
I do not have problems to roll-back it to group and user permissions on the socket. I just want to know if it will look good enough for everyone. @berrange @admiyo ?

**Comment by davidvossel** [comment]
> making sure tenants or other random "users" of the host cannot access libvirt is not a bad idea. We need to solve that not only for the socket but the rest of host resources exposed to the libvirt container. I wonder what are our options...
Either way, it's separate from this PR I guess.

The same thing has to be locked down for docker. This shouldn't be a new problem.

If someone creates a container with the right user permissions that has the docker unix socket mounted, then random users of the host can access the local docker daemon as well.

**Comment by cynepco3hahue** [comment]
@rmohr @davidvossel Guys, can you please review again?

File: cmd/virt-handler/Dockerfile [hunk 15]
```diff
@@ -25,4 +25,10 @@ RUN dnf -y install libvirt-client \
 
 COPY virt-handler /virt-handler
 
+RUN groupadd -g 1000 libvirt && useradd -u 1000 -g 1000 -N --create-home -s /bin/bash virt-handler && \
```
Comment (davidvossel) [inline cmd/virt-handler/Dockerfile]: Do we have to have '--create-home' and '-s' options?

File: cmd/virt-handler/Dockerfile [hunk 16]
```diff
@@ -25,4 +25,10 @@ RUN dnf -y install libvirt-client \
...
+    chown virt-handler /virt-handler
+USER virt-handler
+WORKDIR /home/virt-handler
```
Comment (davidvossel) [inline cmd/virt-handler/Dockerfile]: Would something like this make any more sense?  I'm not sure I understand why the WORKDIR is set to the virt-handler home directory.

```
COPY virt-handler /virt-bin/virt-handler
WORKDIR /virt-bin
RUN groupadd -g 1000 libvirt && useradd -u 1000 -g 1000 -N virt-handler && chown -R virt-handler /virt-bin
USER virt-handler
```

Comment (cynepco3hahue) [reply cmd/virt-handler/Dockerfile]: I believe in this case we will need to change `ENTRYPOINT` under `Dcokerfile`, I prefer to leave virt-handler on the old place. I can omit creating of the home directory and remove `WORKDIR`.

**Comment by cynepco3hahue** [comment]
@rmohr can we re-run Jenkins tests for this one, because I can see that only one test failed on timeout
`Timed out after 30.001s.`

**Comment by cynepco3hahue** [comment]
@davidvossel can you please review again?

**Comment by davidvossel** [comment]
retest this please

**Comment by cynepco3hahue** [comment]
@davidvossel thanks :smile: I can see the light at the end of the tunnel after this patch :fireworks:
Will ask @rmohr to re-run jenkins job, because it has one test that failed on timeout.

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

**Comment by davidvossel** [comment]
retest this please

**Comment by cynepco3hahue** [comment]
@davidvossel @rmohr After last rebase I can see that we mount some host directories
```
volumes:
      - name: libvirt-runtime
        hostPath:
          path: /var/run/libvirt
      - name: sockets
        hostPath:
          path: /var/run/kubevirt
```
hostPath directive automatically inherits host permissions, so if I after user change, virt-handler command can not access these directories, the only option to get access to these directories it to change directory permissions on the host where container runs.
Any thoughts how I can overpower it? Or maybe we just want to leave idea to run virt-handler container as a non-root user?

**Comment by rmohr** [comment]
@cynepco3hahue Yes, it might really make sense to drop this idea for virt-handler. Maybe we will split it later on in more- and less privileged parts, but there are even more changes in the pipeline which might have issues with the switch to a non-privileged user. But in general great work.

**Comment by cynepco3hahue** [comment]
@rmohr ok I will close the pull request and will still save it under my repo :smile:
